controllers/grid/queries/form/QueryForm.php

```diff
@@ -245,135 +251,136 @@ public function fetch($request, $template = null, $display = false, $actionArgs
 
-                // Get current users roles
-                $assignedRoles = [];
-                $usersAssignments = $stageAssignmentDao->getBySubmissionAndStageId($query->getAssocId(), $query->getStageId(), null, $user->getId());
-                while ($usersAssignment = $usersAssignments->next()) {
-                    $userGroup = Repo::userGroup()->get($usersAssignment->getUserGroupId());
-                    $assignedRoles[] = $userGroup->getRoleId();
-                }
+        $stageAssignmentDao = DAORegistry::getDAO('StageAssignmentDAO'); /** @var StageAssignmentDAO $stageAssignmentDao */
 
-                // if current user is editor, add all reviewers
-                if ($user->hasRole([Role::ROLE_ID_SITE_ADMIN], \PKP\core\PKPApplication::CONTEXT_SITE) || $user->hasRole([Role::ROLE_ID_MANAGER], $context->getId()) ||
-                        array_intersect([Role::ROLE_ID_MANAGER, Role::ROLE_ID_SUB_EDITOR], $assignedRoles)) {
-                    foreach ($reviewAssignments as $reviewAssignment) {
-                        $includeUsers[] = $reviewAssignment->getReviewerId();
-                    }
-                }
+        // Get currently selected participants in the query
+        $queryDao = DAORegistry::getDAO('QueryDAO'); /** @var QueryDAO $queryDao */
+        $assignedParticipants = $query->getId() ? $queryDao->getParticipantIds($query->getId()) : [];
+
+        // Always include current user, even if not with a stage assignment
+        $includeUsers[] = $user->getId();
+        $excludeUsers = null;
+
+        // When in review stage, include/exclude users depending on the current users role
+        $reviewAssignments = [];
+        if ($query->getStageId() == WORKFLOW_STAGE_ID_EXTERNAL_REVIEW || $query->getStageId() == WORKFLOW_STAGE_ID_INTERNAL_REVIEW) {
+
+            // Get all review assignments for current submission
+            $reviewAssignmentDao = DAORegistry::getDAO('ReviewAssignmentDAO');
+            $reviewAssignments = $reviewAssignmentDao->getBySubmissionId($submission->getId());
+
+            // Get current users roles
+            $assignedRoles = [];
+            $usersAssignments = $stageAssignmentDao->getBySubmissionAndStageId($query->getAssocId(), $query->getStageId(), null, $user->getId());
+            while ($usersAssignment = $usersAssignments->next()) {
+                $userGroup = Repo::userGroup()->get($usersAssignment->getUserGroupId());
+                $assignedRoles[] = $userGroup->getRoleId();
+            }
 
-                // if current user is an anonymous reviewer, filter out authors
+            // if current user is editor, add all reviewers
+            if ($user->hasRole([Role::ROLE_ID_SITE_ADMIN], \PKP\core\PKPApplication::CONTEXT_SITE) || $user->hasRole([Role::ROLE_ID_MANAGER], $context->getId()) ||
+                    array_intersect([Role::ROLE_ID_MANAGER, Role::ROLE_ID_SUB_EDITOR], $assignedRoles)) {
```

Comment:
I know this isn't your code, but I think the logic here is off. In a case where a journal manager is assigned to a submission in a role with lower permissions, like author, this will show the reviewers to them.

I think the logic needs to be:

```
if (
    has_assignedRoles_manager_or_subeditor 
    OR 
    (has_no_assignedRoles AND is_admin_or_manager)
)
```

Reply:
Also, check the assigned roles first, because `$user->hasRole()` involves a db query. We don't want to make a db query unless we need to.

Reply:
Looks like it is filed here: https://github.com/pkp/pkp-lib/issues/7836. I'll add a note there and you can leave this as-is for now.